concepts/websockets.textile

```diff
@@ -0,0 +1,591 @@
+
+h4(#minimal-code-samples-1). Minimal code samples
+
+*Client (Browser, before bundling):*
+
+```[js]
+var WebSocket = require('faye-websocket'),
+    ws        = new WebSocket.Client('ws://www.example.com/');
+
+ws.on('open', function(event) {
+  console.log('open');
+  ws.send('Hello, world!');
+});
+
+ws.on('message', function(event) {
+  console.log('message', event.data);
+});
+
+ws.on('close', function(event) {
+  console.log('close', event.code, event.reason);
+  ws = null;
+});
+```
+
+*Server (Node.js):*
+
+```[js]
+var WebSocket = require('faye-websocket'),
+    http      = require('http');
+
+var server = http.createServer();
+
+server.on('upgrade', function(request, socket, body) {
+  if (WebSocket.isWebSocket(request)) {
+    var ws = new WebSocket(request, socket, body);
+    
+    ws.on('message', function(event) {
+      ws.send(event.data);
+    });
+    
+    ws.on('close', function(event) {
+      console.log('close', event.code, event.reason);
+      ws = null;
+    });
+  }
+});
+
+server.listen(8000);
+```
+
+h3(#socket.io). Socket.io
+
+"Socket.io":https://github.com/socketio/socket.io has been around for a while now; I like to think of it as the "jQuery" of WebSockets. It uses long polling and WebSockets for its transports, by default starting with long polling and then upgrading to WebSockets if available. Given the waning relevance of long polling, Socket.io's main drawcards these days are its "other features":https://stackoverflow.com/questions/38546496/moving-from-socket-io-to-raw-websockets/38546537#38546537, such as restoring dropped connections, automatic support for JSON, and "namespaces", which are essentially isolated messaging channels multiplexed over the same client connection.
+
+Socket.io isn't actually interchangeable with generic WebSockets solutions - either on the server side or on the client side - and attempting to connect to something other than a Socket.io client or server will fail. It has its own additional handshake protocol, and some additional metadata included in each message.
+
+Should you use it? On the plus side, in addition to "the simplicity of getting up and running":https://socket.io/get-started/chat/, it's well-established and there is a wealth of learning material out there if you get stuck. On the other hand, there have been plenty of reports of memory leaks and general performance issues over time, and despite there being "a mountain of issues":https://github.com/socketio/socket.io/issues in the Github repository, responses are few, and "commits and updates":https://github.com/socketio/socket.io/commits/master these days seem relatively infrequent. Just like jQuery, I think Socket.io is a product of a bygone era, and for new projects you'd be better off with something a little more modern. jQuery does still have its fans though, so...
```

Comment:
The sentence following the question 'Should you use it?' is irrelevant so perhaps just skip the question.

Reply:
'Just like jQuery, I think Socket.io is a ' avoid using 'I' please.

Reply:
- Not sure I agree with the first request - the sentences that follow are answers to the question and lose applicability if the question is removed.
- Left over from the article's origins as a blog post. I have reworded this.